libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayHistorySession.kt

```diff
@@ -0,0 +1,115 @@
+import com.mapbox.navigation.utils.internal.logI
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.update
+
+/**
+ * Used to create a replay trip session.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class ReplayHistorySession : MapboxNavigationObserver {
+
+    private val optionsFlow = MutableStateFlow(
+        ReplayHistorySessionOptions.Builder().build()
+    )
+    private var mapboxHistoryReader: MapboxHistoryReader? = null
+    private var mapboxNavigation: MapboxNavigation? = null
+    private var lastHistoryEvent: ReplayEventBase? = null
+    private var coroutineScope: CoroutineScope? = null
+
+    private val replayEventsObserver = ReplayEventsObserver { events ->
+        logI(LOG_CATEGORY) { "replayEvents ${events.joinToString("\n")}" }
+        if (isLastEventPlayed(events)) {
+            pushMorePoints()
+        }
+    }
+
+    private fun isLastEventPlayed(events: List<ReplayEventBase>): Boolean {
+        val currentLocationEvent = events.lastOrNull { it is ReplayEventUpdateLocation }
+            ?: return false
+        val lastEventTimestamp = this.lastHistoryEvent?.eventTimestamp ?: 0.0
+        return currentLocationEvent.eventTimestamp >= lastEventTimestamp
+    }
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
+            .also { this.coroutineScope = it }
+        this.mapboxNavigation = mapboxNavigation
+        this.lastHistoryEvent = null
+        mapboxNavigation.startReplayTripSession()
+        observeStateFlow(mapboxNavigation).launchIn(coroutineScope)
+        mapboxNavigation.mapboxReplayer.registerObserver(replayEventsObserver)
+    }
+
+    inline fun <T, R> Flow<T>.mapDistinct(
+        crossinline transform: suspend (value: T) -> R
+    ): Flow<R> = map(transform).distinctUntilChanged()
+
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        coroutineScope?.cancel()
+        coroutineScope = null
+        this.mapboxNavigation = null
+        mapboxNavigation.mapboxReplayer.unregisterObserver(replayEventsObserver)
+        mapboxNavigation.mapboxReplayer.stop()
+        mapboxNavigation.mapboxReplayer.clearEvents()
+    }
+
+    fun getOptions(): StateFlow<ReplayHistorySessionOptions> = optionsFlow.asStateFlow()
+
+    fun setOptions(options: ReplayHistorySessionOptions) = optionsFlow.update { options }
+
+    fun setHistoryFile(absolutePath: String) {
```

Comment:
Does it work fine when you set the file in the middle of replaying another one? Maybe `historyFile` should be a constructor parameter.

Reply:
> Does it work fine when you set the file in the middle of replaying another one

This is what I'm adding support for. This is what the ReplayActivity does, you select history files from a list. The ReplayHistorySession ensures the trip session is reset and the new location can be started.

The other option is to move that responsibility to `ReplayTripStarter`. I'm seeing it as the same result. People preferred "functions over options" in the ReplayTripStarter pull request, so I was using functions here.